spectec/spec/7-module.watsup

```diff
@@ -0,0 +1,134 @@
+;; Allocation
+
+
+def $funcs(externval*) : funcaddr*
+def $globals(externval*) : globaladdr*
+def $tables(externval*) : tableaddr*
+def $mems(externval*) : memaddr*
+
+def $funcs(externval*) = [fa|FUNC fa <- externval]
+def $globals(externval*) = [ga|GLOBAL ga <- externval]
+def $tables(externval*) = [ta|TABLE ta <- externval]
+def $mems(externval*) = [ma|MEM ma <- externval]
+
+
+def $instexport(funcaddr*, globaladdr*, tableaddr*, memaddr*, export) : exportinst
+def $instexport(fa*, ga*, ta*, ma*, EXPORT name (FUNC x)) = { NAME name, VALUE (FUNC fa*[x]) }
+def $instexport(fa*, ga*, ta*, ma*, EXPORT name (GLOBAL x)) = { NAME name, VALUE (GLOBAL ga*[x]) }
+def $instexport(fa*, ga*, ta*, ma*, EXPORT name (TABLE x)) = { NAME name, VALUE (TABLE ta*[x]) }
+def $instexport(fa*, ga*, ta*, ma*, EXPORT name (MEM x)) = { NAME name, VALUE (MEM ma*[x]) }
+
+
+def $allocfunc(moduleinst, func) : funcinst
+def $allocfunc(m, func) = { MODULE m, CODE func }
+
+def $allocglobal(globaltype, val) : globalinst
+def $allocglobal(globaltype, val) = { TYPE globaltype, VALUE val }
+
+def $alloctable(tabletype) : tableinst
+def $alloctable(`[i .. j] rt) = { TYPE (`[i .. j] rt), ELEM (REF.NULL rt)^i }
+
+def $allocmem(memtype) : meminst
+def $allocmem(`[i .. j] I8) = { TYPE (`[i .. j] I8), DATA 0^(i * 64 * $Ki()) }
+
+def $allocelem(reftype, ref*) : eleminst
+def $allocelem(rt, ref*) = { TYPE rt, ELEM ref* }
+
+def $allocdata(byte*) : datainst
+def $allocdata(byte*) = { DATA byte* }
+
+
+def $allocmodule(store, module, externval*, val*, (ref*)*) : (store, moduleinst)
+def $allocmodule(s, module, externval*, val*, (ref*)*) = (s_6, m)
+  -- if module =
+    MODULE
+      import*
+      func*
+      (GLOBAL globaltype expr)*
+      (TABLE tabletype)*
+      (MEMORY memtype)*
+      (ELEM rt expr* elemmode?)*
+      (DATA byte* datamode?)*
+      start?
+      export*
+  -- if fa_ex* = $funcs(externval*)
+  -- if ga_ex* = $globals(externval*)
+  -- if ta_ex* = $tables(externval*)
+  -- if ma_ex = $mems(externval*)
+  -- if fa* = |s.FUNC| ... $(|s.FUNC| + |func*| - 1)
+  -- if ga* = |s.GLOBAL| ... $(|s.GLOBAL| + |global*| - 1)
+  -- if ta* = |s.TABLE| ... $(|s.TABLE| + |table*| - 1)
+  -- if ma* = |s.MEM| ... $(|s.MEM| + |mem*| - 1)
+  -- if ea* = |s.ELEM| ... $(|s.ELEM| + |elem*| - 1)
+  -- if da* = |s.DATA| ... $(|s.DATA| + |data*| - 1)
```

Comment:
I suggest new syntax for range to precomputing the addresses.

Reply:
Can this perhaps be expressed by an auxiliary function?
```
def $range(nat_1, nat_2) : nat*  hint(show %...%)
def $range(n_1, n_2) = epsilon  -- if n_1 > n_2
def $range(n_1, n_2) = n_1 $range(n_1+1, n_2)  -- otherwise
```

Reply:
Thanks for the idea. I'll reflect it.